plugins/image-prioritizer/class-image-prioritizer-img-tag-visitor.php

```diff
@@ -144,36 +188,72 @@ public function __invoke( OD_Tag_Visitor_Context $context ): bool {
 
 		// If this element is the LCP (for a breakpoint group), add a preload link for it.
 		foreach ( $context->url_metric_group_collection->get_groups_by_lcp_element( $xpath ) as $group ) {
-			$link_attributes = array_merge(
-				array(
-					'rel'           => 'preload',
-					'fetchpriority' => 'high',
-					'as'            => 'image',
-				),
-				array_filter(
+			if ( $this->is_within_picture && count( $this->collected_sources ) > 0 ) {
+				foreach ( $this->collected_sources as $source ) {
+					$link_attributes = array_merge(
+						array(
+							'rel'           => 'preload',
+							'fetchpriority' => 'high',
+							'as'            => 'image',
+						),
+						array_filter(
+							array(
+								'href'        => isset( $source['srcset'] ) && is_string( $source['srcset'] )
+													? explode( ' ', $source['srcset'] )[0]
+													: '',
+								'imagesrcset' => isset( $source['srcset'] ) && is_string( $source['srcset'] ) ? $source['srcset'] : '',
+								'imagesizes'  => isset( $source['sizes'] ) && is_string( $source['sizes'] ) ? $source['sizes'] : '',
+								'type'        => isset( $source['type'] ) && is_string( $source['type'] ) ? $source['type'] : '',
+								'media'       => isset( $source['media'] ) && is_string( $source['media'] ) ? 'screen and ' . $source['media'] : 'screen',
+							),
+							static function ( string $value ): bool {
+								return '' !== $value;
+							}
+						)
+					);
+
+					if ( isset( $source['crossorigin'] ) ) {
+						$link_attributes['crossorigin'] = 'use-credentials' === $source['crossorigin'] ? 'use-credentials' : 'anonymous';
+					}
+
+					$context->link_collection->add_link(
+						$link_attributes,
+						$group->get_minimum_viewport_width(),
+						$group->get_maximum_viewport_width()
+					);
+				}
+			} else {
```

Comment:
As noted above, I think this logic should rather go into a `process_picture()` method along with the other new logic above.